File: src_py/tools/info/get_api_info.py
"""Get API key information and organization details from Dixa"""
import os
import json
from tools.utils import get_api_key, make_request


def mask_api_key(api_key: str) -> str:
    """
    Mask an API key showing only the first 4 and last 4 characters.
    
    Args:
        api_key: The API key to mask
    
    Returns:
        Masked API key string (e.g., "dixa_xxxx...xxxx")
    """
    if len(api_key) <= 8:
        return "*" * len(api_key)
    
    return f"{api_key[:4]}...{api_key[-4:]}"


def get_api_info(log=None, session=None) -> str:
    """
    Get information about the configured Dixa API key and the associated organization.
    
    This tool shows:
    - A masked version of the API key (first 4 and last 4 characters)
    - Organization information from Dixa API
    
    Args:
        log: Optional logger for debugging
        session: Optional session context that may contain auth information
    
    Returns:
        JSON string containing API key info and organization details
    """
    try:
        api_key = get_api_key(session)
        masked_key = mask_api_key(api_key)
        
        if log:
            log.debug("Fetching organization information")
        
        # Try to get organization info from Dixa API
        # Common endpoints: /v1/organization, /v1/organizations, /v1/me
        organization_info = None
        error_message = None
        
        # Try /v1/organization endpoint first
        try:
            url = "https://dev.dixa.io/v1/organization"
            org_data = make_request("GET", url, log=log, session=session)
            organization_info = json.loads(org_data)
            # Handle case where response is wrapped in 'data' key
            if isinstance(organization_info, dict) and "data" in organization_info:
                organization_info = organization_info["data"]
        except Exception as e:
            error_message = str(e)
            # Try alternative endpoint /v1/organizations
            try:
                url = "https://dev.dixa.io/v1/organizations"
                org_data = make_request("GET", url, log=log, session=session)
                organization_info = json.loads(org_data)
                # Handle case where response is wrapped in 'data' key
                if isinstance(organization_info, dict) and "data" in organization_info:
                    organization_info = organization_info["data"]
                error_message = None
            except Exception:
                pass
        
        result = {
            "api_key": {
                "masked": masked_key,
                "length": len(api_key),
                "is_set": True
            },
            "organization": organization_info if organization_info else None,
            "error": error_message if error_message else None
        }
        
        return json.dumps(result, indent=2)
        
    except ValueError as e:
        # API key not set
        return json.dumps({
            "api_key": {
                "masked": "NOT SET",
                "length": 0,
                "is_set": False
            },
            "organization": None,
            "error": str(e)
        }, indent=2)
    except Exception as e:
        return json.dumps({
            "api_key": {
                "masked": "ERROR",
                "length": 0,
                "is_set": False
            },
            "organization": None,
            "error": str(e)
        }, indent=2)

File: src/tools/info/getApiInfo.ts
import { z } from "zod";
import { getDixaApiKey } from "../utils";

/**
 * Mask an API key showing only the first 4 and last 4 characters.
 */
function maskApiKey(apiKey: string): string {
  if (apiKey.length <= 8) {
    return "*".repeat(apiKey.length);
  }
  return `${apiKey.substring(0, 4)}...${apiKey.substring(apiKey.length - 4)}`;
}

export const getApiInfo = {
  name: "getApiInfo",
  description: "Preview the configured DIXA_API_KEY (masked) and get information about the associated organization",
  parameters: z.object({}),
  execute: async (args, { log, session }) => {
    let apiKey: string | null = null;
    let maskedKey = "NOT SET";
    let apiKeyLength = 0;
    let isSet = false;

    try {
      apiKey = getDixaApiKey(session);
      maskedKey = maskApiKey(apiKey);
      apiKeyLength = apiKey.length;
      isSet = true;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return {
        type: "text" as const,
        text: JSON.stringify({
          api_key: {
            masked: "NOT SET",
            length: 0,
            is_set: false,
          },
          organization: null,
          error: errorMessage,
        }, null, 2),
      };
    }

    // Try to get organization info from Dixa API
    let organizationInfo: any = null;
    let errorMessage: string | null = null;

    // Try /v1/organization endpoint first
    try {
      const url = "https://dev.dixa.io/v1/organization";
      log.debug("Request URL:", url);

      const resp = await fetch(url, {
        method: "GET",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
      });

      if (resp.ok) {
        const data = await resp.json();
        // Handle case where response is wrapped in 'data' key
        organizationInfo = data?.data || data;
      } else {
        const errorText = await resp.text();
        errorMessage = `Failed to fetch organization: ${resp.status} ${resp.statusText}\nResponse: ${errorText}`;
        
        // Try alternative endpoint /v1/organizations
        try {
          const altUrl = "https://dev.dixa.io/v1/organizations";
          log.debug("Trying alternative URL:", altUrl);

          const altResp = await fetch(altUrl, {
            method: "GET",
            headers: {
              Authorization: `Bearer ${apiKey}`,
              "Content-Type": "application/json",
            },
          });

          if (altResp.ok) {
            const altData = await altResp.json();
            organizationInfo = altData?.data || altData;
            errorMessage = null;
          } else {
            const altErrorText = await altResp.text();
            errorMessage = `Failed to fetch organization from both endpoints. Last error: ${altResp.status} ${altResp.statusText}\nResponse: ${altErrorText}`;
          }
        } catch (altError) {
          // Keep the original error message
        }
      }
    } catch (error) {
      errorMessage = error instanceof Error ? error.message : String(error);
    }

    const result = {
      api_key: {
        masked: maskedKey,
        length: apiKeyLength,
        is_set: isSet,
      },
      organization: organizationInfo,
      error: errorMessage,
    };

    return {
      type: "text" as const,
      text: JSON.stringify(result, null, 2),
    };
  },
};

